### PR TITLE
Update to cats-effect 1.0.0

### PR DESCRIPTION
**Pull request by joroKr21**
Including some additional tweaks:

* Implement `asyncF` and `bracketCase`
* Reorder methods a bit
* Override `map` and `toIO` for performance
* Factor out `registerCallback`
* Update scala and scalaz

### CODE REVIEW

**Comment by joroKr21** [comment]
Ref #7

**Comment by joroKr21** [comment]
Hmm, looks like `toIO` is not stack-safe, but it's implemented with `runAsync` and I didn't change it.

File: core/src/main/scala/io/chrisdavenport/scalaz/task/instances/TaskInstances.scala [hunk 1]
```diff
@@ -37,27 +50,25 @@ trait TaskInstances {
...
       **/
-    def runAsync[A](fa: Task[A])(cb: Either[Throwable, A] => IO[Unit]): IO[Unit] =
-      IO(
-        fa.unsafePerformAsync { disjunction =>
-          cb(disjunction.toEither)
-            .unsafeRunAsync(_ => ())
-        }
-      )
+    def runAsync[A](fa: Task[A])(cb: Either[Throwable, A] => IO[Unit]): SyncIO[Unit] =
+      SyncIO(cb(fa.unsafePerformSyncAttempt.toEither).unsafeRunAsyncAndForget())
```
Comment (joroKr21) [inline core/src/main/scala/io/chrisdavenport/scalaz/task/instances/TaskInstances.scala]: I'm not sure if this implementation makes any sense but it does make the tests pass.

Comment (joroKr21) [reply core/src/main/scala/io/chrisdavenport/scalaz/task/instances/TaskInstances.scala]: No, this implementation is not right, it shouldn't be blocking. But I can't seem to implement this method in any way that doesn't spuriously fail.

**Comment by joroKr21** [comment]
Ah, I think I get it now. `Arbitrary` for `Task` is using the default `ExecutorService`

**Comment by joroKr21** [comment]
@ChristopherDavenport this is ready for review. At this point it's the only dependency blocking me from upgrading to cats-effect 1.0. Wdyt?

**Review by ChristopherDavenport — APPROVED** [review]

**Comment by ChristopherDavenport** [comment]
@joroKr21  Released as RC1. If everything looks good for you in a week we can cut the full release.

**Comment by joroKr21** [comment]
I finally managed to upgrade which was a bit more tricky than I thought originally. But I still needed to implement `ContextShift[Task]`. Do you think it would make a good addition?

**Comment by ChristopherDavenport** [comment]
I am happy to put lawful constructs in place.

**Comment by joroKr21** [comment]
I don't think `ContextShift` has any laws. It's explicitly documented as "not a typeclass". It plays together with `Concurrent` but `Concurrent[Task]` doesn't seem possible. Anyway, `ContextShift` is just a few lines of code so anyone could implement it, but it uses a bit of scalaz internals which people might not know about.

**Pull request merged**